src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -0,0 +1,235 @@
+
+                processed += batchSize;
+            }
+        }
+
+        logger.LogInformation("Completed updating all instance blobs");
+    }
+
+    /// <summary>
+    /// Asynchronously commits all the instances in a study and creates new entries for changefeed.
+    /// </summary>
+    /// <param name="arguments">CompleteInstanceArguments</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// A task representing the <see cref="CompleteUpdateInstanceAsync"/> operation.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="arguments"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(CompleteUpdateInstanceAsync))]
+    public async Task CompleteUpdateInstanceAsync([ActivityTrigger] CompleteInstanceArguments arguments, ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(arguments.ChangeDataset, nameof(arguments.ChangeDataset));
+        EnsureArg.IsNotNull(arguments.StudyInstanceUid, nameof(arguments.StudyInstanceUid));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        logger.LogInformation("Completing updating operation for study.");
+
+        try
+        {
+            await _indexStore.EndUpdateInstanceAsync(
+                arguments.PartitionKey,
+                arguments.StudyInstanceUid,
+                GetDeserialzedDataset(arguments.ChangeDataset),
+                CancellationToken.None);
+
+            logger.LogInformation("Updating study completed successfully.");
+        }
+        catch (StudyNotFoundException)
+        {
+            // TODO: Study deleted, we need to cleanup all the newly updated blobs.
+            logger.LogWarning("Failed to update to study. Possibly deleted.");
+        }
+    }
+
+    /// <summary>
+    /// Asynchronously delete all the old blobs if it has more than 2 version.
+    /// </summary>
+    /// <param name="context">Activity context which has list of watermarks to cleanup</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// A task representing the <see cref="DeleteOldVersionBlobAsync"/> operation.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="context"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(DeleteOldVersionBlobAsync))]
+    public async Task DeleteOldVersionBlobAsync([ActivityTrigger] IDurableActivityContext context, ILogger logger)
+    {
+        EnsureArg.IsNotNull(context, nameof(context));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        IReadOnlyList<InstanceFileIdentifier> fileIdentifiers = context.GetInput<IReadOnlyList<InstanceFileIdentifier>>();
+
+        logger.LogInformation("Begin deleting old blobs. Total size {TotalCount}", fileIdentifiers.Count);
+
+        await Parallel.ForEachAsync(
```

Comment:
nit: you can filter the identifiers to avoid unnecessary threads:
```csharp
        await Parallel.ForEachAsync(
            fileIdentifiers.Where(f => f.OriginalVersion.HasValue),
            new ParallelOptions
            {
                CancellationToken = default,
                MaxDegreeOfParallelism = _options.MaxParallelThreads,
            },
            (fileIdentifier, token) => _updateInstanceService.DeleteInstanceBlobAsync(fileIdentifier.Version, token),
            });
```

Reply:
Good catch!!